proposed/http-middleware/middleware.md

```diff
@@ -1,114 +1,123 @@
 ======================
 
-This document describes a common standard for HTTP server middleware components
-using HTTP Messages defined by [PSR-7](http://www.php-fig.org/psr/psr-7/).
+This document describes common interfaces for HTTP server middleware components
+("middleware") and HTTP server request handlers ("handlers") that use HTTP
+messages as described by [PSR-7][psr7].
 
-HTTP middleware has been an important concept on other web development platforms
-for a good number of years, and since the introduction of a formal HTTP Messages
-standard has been growing increasingly popular with web frameworks.
+HTTP middleware is a way to move common request and response processing away from
+the application layer. HTTP request handlers are a fundamental part of any web
+application. Server side code receives an HTTP request message, processes it, and
+then produces a response.
 
 The interfaces described in this document are abstractions for HTTP middleware
-and the containers that are used to process HTTP requests.
+and HTTP request handlers.
 
-_**Note:** Any references to "middleware" in this document are specific to
-**server middleware**._
+_Note: All references to "handlers" and "middleware" are specific to **server
+request** processing._
 
 The key words "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT", "SHOULD",
 "SHOULD NOT", "RECOMMENDED", "MAY", and "OPTIONAL" in this document are to be
-interpreted as described in [RFC 2119](http://tools.ietf.org/html/rfc2119).
+interpreted as described in [RFC 2119][rfc2119].
+
+[psr7]: http://www.php-fig.org/psr/psr-7/
+[rfc2119]: http://tools.ietf.org/html/rfc2119
 
 ### References
 
-- [RFC 2119](http://tools.ietf.org/html/rfc2119)
-- [PSR-7](http://www.php-fig.org/psr/psr-7/)
+- [PSR-7][psr7]
+- [RFC 2119][rfc2119]
 
 ## 1. Specification
 
-An HTTP middleware component is an individual component participating, together
-with other middleware components, in the processing of an incoming HTTP Request
-and the creation of a resulting HTTP Response, as defined by PSR-7.
+### 1.1 Middleware
 
-Middleware using this standard MUST implement the following interface:
+An HTTP middleware component is an individual component participating, often
+together with other middleware components, in the processing of an incoming
+request and the creation of a resulting response, as defined by PSR-7.
 
-- `Psr\Http\ServerMiddleware\MiddlewareInterface`
+A middleware component MAY create and return a response without delegating to
+a request handler, if sufficient conditions are met.
 
-Middleware dispatching systems using this standard MUST implement the following
-interface:
+Middleware using this standard MUST implement the following interface:
 
-- `Psr\Http\ServerMiddleware\DelegateInterface`
+- `Psr\Http\Server\MiddlewareInterface`
 
-Legacy middleware implementing a double pass approach MUST be wrapped using an
-object that implements the `MiddlewareInterface`.
+### 1.2 Request Handlers
```

Comment:
I think we need to invert sections 1.1 and 1.2, as the discussion of middleware relies on request handlers.